com/williamfiset/algorithms/dp/VirtualMachinesScheduler.java

```diff
@@ -0,0 +1,177 @@
+ * cost of the complex process until step i when it runs on VM-j. In order for step i to run on
+ * VM-j, the results from i-1 step must already be on VM-j or sent on VM-j including the
+ * communication cost between VMs.
+ *
+ * <p>For the example above the Cost array is the following:
+ *
+ * <p>5 6 3 12 13 8 17 18 11 15 20 17
+ *
+ * <p>This means that the least expensive execution costs 15 and is achieved when the last process
+ * is on the first VM.
+ *
+ * @author Dragatis Nikolas, nikosdraga@gmail.com
+ */
+public class VirtualMachinesScheduler {
+  public static void main(String[] args) throws Exception {
+    double time = System.currentTimeMillis();
+    Scheduler scheduler = new Scheduler();
+
+    scheduler.scanFile(args[0]);
+
+    System.out.println("\nTime: " + (System.currentTimeMillis() - time) / 1000.0 + " seconds.");
+  }
+
+  /**
+   * This is the class that contains the methods for producing the final array of the dynamic
+   * programming after reading and storing the data from a txt file.
+   */
+  public static class Scheduler {
+    private int virtualMachines = 0;
+    private int processes = 0;
+
+    /**
+     * This function is used to scan the txt file, with the data, and initialize the dimensions of
+     * the array with the processes costs and the array with the communication costs.
+     *
+     * @param pathname is the name of the txt file containing the data.
+     * @throws Exception
+     */
+    public void scanFile(String pathname) throws Exception {
+      Scanner scanner = new Scanner(new FileReader(pathname));
+
+      processes = scanner.nextInt();
+      virtualMachines = scanner.nextInt();
+
+      int[][] processesCost = new int[processes][virtualMachines];
+      int[][] communicationsCost = new int[virtualMachines][virtualMachines];
+
+      initializeArrays(scanner, processesCost, processes, virtualMachines);
+
+      initializeArrays(scanner, communicationsCost, virtualMachines, virtualMachines);
+
+      Scheduling(processesCost, communicationsCost);
+    }
+
+    /**
+     * This function is used to produce the final array of the dynamic programming with the results.
+     *
+     * @param processesCost is the array which contains each process cost on each virtual machine.
+     * @param communicationsCost is the array which contains the costs to transfer data from one
+     *     virtual machine to another.
+     */
+    public void Scheduling(int[][] processesCost, int[][] communicationsCost) {
+      int[][] Cost = new int[processes][virtualMachines];
+
+      for (int i = 0; i < virtualMachines; i++) {
+        Cost[0][i] = processesCost[0][i];
+      }
+
+      for (int i = 1; i < processes; i++) {
+        for (int j = 0; j < virtualMachines; j++) {
+          Cost[i][j] = findMin(i, j, Cost, processesCost, communicationsCost, virtualMachines);
+        }
+      }
+
+      printArray(Cost, processes, virtualMachines);
```

Comment:
Instead of printing, this should return a value?

Reply:
It could return the Cost array and use the printArray() function in scanFile(), but I think it's the same thing.

Reply:
It's printing the result. But you are right, you can't use a tester to verify the results. So, should it return the Cost array in order to test the results and then print it?